lib/policies/proxy/proxy.js

```diff
@@ -12,73 +12,67 @@ const createStrategy = (strategy, proxyOptions, endpointUrls) => {
-      'policy configuration) does not exist');
+  if (!endpoint) { // Note: one day this can be ensured by JSON Schema, when $data keyword will be avaiable.
+    throw new Error(`service endpoint ${serviceEndpointKey} (referenced in proxy policy configuration) does not exist`);
   }
 
-  if (!endpoint.url && !endpoint.urls) {
-    throw new Error(
-      `service endpoint ${serviceEndpointKey} (referenced in 'proxy' ` +
-      'policy configuration) does not contain a `url` or `urls` property');
+  const proxyOptions = parseProxyOptions(params.proxyOptions, endpoint);
+  const intermediateProxyUrl = process.env.HTTP_PROXY || params.proxyUrl;
+
+  if (intermediateProxyUrl) {
+    logger.info(`using intermediate proxy ${intermediateProxyUrl}`);
+    proxyOptions.agent = new ProxyAgent(intermediateProxyUrl);
   }
 
-  const proxy = httpProxy.createProxyServer({ changeOrigin: params.changeOrigin });
+  const proxy = httpProxy.createProxyServer(Object.assign({ changeOrigin: params.changeOrigin }, proxyOptions));
 
-  proxy.on('error', (err, _req, res) => {
+  proxy.on('error', (err, req, res) => {
     logger.warn(err);
 
-    if (!res) {
-      throw err;
-    }
-
     if (!res.headersSent) {
       res.status(502).send('Bad gateway.');
     } else {
       res.end();
     }
   });
 
-  let balancer;
+  let strategy;
+  let urls;
+
   if (endpoint.url) {
-    // single `url` property takes precedence over `urls` array
-    balancer = createStrategy('static', proxyOptions, [endpoint.url]);
+    strategy = 'static';
+    urls = [endpoint.url];
   } else {
-    const strategy = params.strategy || 'round-robin';
-    balancer = createStrategy(strategy, proxyOptions, endpoint.urls);
+    strategy = params.strategy || 'round-robin';
+    urls = endpoint.urls;
   }
 
-  const intermediateProxySettings = process.env.http_proxy || params.proxyUrl;
-  if (intermediateProxySettings) {
-    logger.debug(`using intermediate proxy ${intermediateProxySettings}`);
-    proxyOptions.agent = new ProxyAgent(intermediateProxySettings);
-  }
+  const balancer = createStrategy(strategy, proxyOptions, urls);
 
   return function proxyHandler (req, res) {
-    const targetUrl = balancer.nextTarget();
-    proxyOptions.target = Object.assign(proxyOptions.target, targetUrl);
+    const target = balancer.nextTarget();
+    const headers = prepareHeaders(req.egContext);
 
-    logger.debug(`proxying to ${targetUrl.href}, ${req.method} ${req.url}`);
+    logger.debug(`proxying to ${target.href}, ${req.method} ${req.url}`);
 
-    prepareHeaders(params, proxyOptions, req.egContext);
-    proxy.web(req, res, proxyOptions);
+    proxy.web(req, res, { target, headers });
   };
 
   // multiple urls will load balance, defaulting to round-robin
 };
 
-function prepareHeaders (params, proxyOptions, egContext) {
-  proxyOptions.headers = proxyOptions.headers || {};
+function prepareHeaders (egContext) {
+  const headers = {};
```

Comment:
what if I want some custom predefined headers? `proxyOptions.headers`

Reply:
you can define some custom static headers via `proxyOptions.headers` option. It looks like it is no longer supported

Reply:
You're right. Great catch! I'll fix this right now and maybe add a test as well for this.